File: src/main/java/ru/arc/audit/AuditManager.java
package ru.arc.audit;

import ru.arc.ARC;
import ru.arc.configs.Config;
import ru.arc.configs.ConfigManager;
import ru.arc.network.repos.RedisRepo;
import ru.arc.util.Utils;
import ru.arc.xserver.playerlist.PlayerManager;
import lombok.extern.slf4j.Slf4j;
import net.kyori.adventure.audience.Audience;
import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.ConcurrentLinkedDeque;

import static ru.arc.util.Logging.*;
import static ru.arc.util.TextUtil.mm;
import static java.nio.file.StandardOpenOption.*;

@Slf4j
public class AuditManager {

    private static final Config config = ConfigManager.of(ARC.plugin.getDataPath(), "audit.yml");
    private static RedisRepo<AuditData> repo;

    private static BukkitTask pruneTask;
    private static BukkitTask balanceHistoryTask;

    private static final Path balanceHistoryPath = ARC.plugin.getDataPath().resolve("balance-history");

    public static void init() {
        cancel();
        if (!Files.exists(balanceHistoryPath)) {
            try {
                Files.createDirectories(balanceHistoryPath);
            } catch (IOException e) {
                error("Failed to create balance history directory", e);
            }
        }
        if (repo == null) {
            repo = RedisRepo.builder(AuditData.class)
                    .saveBackups(false)
                    .id("audit")
                    .redisManager(ARC.redisManager)
                    .storageKey("arc.audits")
                    .saveInterval((long) config.integer("save-interval", 20))
                    .updateChannel("arc.audit-update")
                    .loadAll(false)
                    .build();
        }

        pruneTask = new BukkitRunnable() {
            @Override
            public void run() {
                long maxAge = config.integer("max-age-seconds", 86400 * 30) * 1000L;
                int maxWeight = config.integer("max-weight", 100000);
                long weight = weight();
                int count = 0;
                while (weight > maxWeight) {
                    info("Pruning audit data, weight: {}, maxAge: {}", weight, maxAge);
                    for (AuditData auditData : repo.all()) {
                        auditData.trim(maxAge);
                    }
                    weight = weight();
                    maxAge /= 2;
                    count++;
                    if (count > 10) {
                        warn("Pruning audit data failed to reduce weight below maxWeight: {}", maxWeight);
                        break;
                    }
                }
            }
        }.runTaskTimerAsynchronously(ARC.plugin,
                config.integer("prune-interval", 6000),
                config.integer("prune-interval", 6000));

        balanceHistoryTask = new BukkitRunnable() {
            @Override
            public void run() {
                if (!config.bool("balance-history", false)) {
                    return;
                }
                for (String player : PlayerManager.getPlayerNames()) {
                    OfflinePlayer offlinePlayer = Bukkit.getOfflinePlayer(player);
                    if (offlinePlayer.getName() == null || !offlinePlayer.getName().equalsIgnoreCase(player)) {
                        error("Failed to get offline player for {}", player);
                        continue;
                    }
                    double balance = ARC.getEcon().getBalance(offlinePlayer);
                    Path playerPath = balanceHistoryPath.resolve(player + ".csv");
                    long timestamp = System.currentTimeMillis();
                    try {
                        if (!Files.exists(playerPath)) {
                            Files.createFile(playerPath);
                        }
                        Files.write(playerPath, (timestamp + "," + balance + "\n").getBytes(), CREATE, APPEND, WRITE);
                        //info("Saved balance data for {}", player);
                    } catch (IOException e) {
                        error("Failed to write balance history for {}", player, e);
                    }
                }
            }
        }.runTaskTimerAsynchronously(ARC.plugin, 20, 20L * 60 * 5);
    }

    public static void join(String name) {
        repo.addContext(name.toLowerCase());
    }

    public static void leave(String name) {
        repo.removeContext(name.toLowerCase());
    }


    public static void cancel() {
        if (pruneTask != null && !pruneTask.isCancelled()) {
            pruneTask.cancel();
        }
        if (balanceHistoryTask != null && !balanceHistoryTask.isCancelled()) {
            balanceHistoryTask.cancel();
        }
    }

    public static void operation(String name, double amount, Type type, String comment) {
        repo.getOrCreate(name.toLowerCase(), () -> new AuditData(new ConcurrentLinkedDeque<>(), name, System.currentTimeMillis()))
                .thenAccept(auditData -> auditData.operation(amount, type, comment));
    }

    public static long weight() {
        long weight = 0;
        for (AuditData auditData : repo.all()) {
            weight += auditData.getTransactions().size();
        }
        return weight;
    }

    public static void sendAudit(Audience player, String auditedPlayer, int page, Filter filter) {
        repo.getOrCreate(auditedPlayer.toLowerCase(), () -> new AuditData(new ConcurrentLinkedDeque<>(), auditedPlayer, System.currentTimeMillis()))
                .thenAccept(auditData -> {
                    if (auditData == null) {
                        player.sendMessage(config.componentDef("messages.no-audit-data", "<red>No audit data found for %player_name%",
                                "%player_name%", auditedPlayer));
                        return;
                    }
                    List<Transaction> transactions = new ArrayList<>(auditData.getTransactions());
                    if (filter != null && filter != Filter.ALL) {
                        if (filter == Filter.INCOME) {
                            transactions.removeIf(transaction -> transaction.getAmount() < 0);
                        } else if (filter == Filter.EXPENSE) {
                            transactions.removeIf(transaction -> transaction.getAmount() > 0);
                        } else if (filter == Filter.SHOP) {
                            transactions.removeIf(transaction -> transaction.getType() != Type.SHOP);
                        } else if (filter == Filter.JOB) {
                            transactions.removeIf(transaction -> transaction.getType() != Type.JOB);
                        } else if (filter == Filter.PAY) {
                            transactions.removeIf(transaction -> transaction.getType() != Type.PAY);
                        }
                    }
                    player.sendMessage(formatAudit(transactions.reversed(), auditedPlayer, page, filter));
                });
    }

    public enum Filter {
        INCOME, EXPENSE, ALL, SHOP, JOB, PAY
    }

    private static Component formatAudit(Collection<Transaction> reversedAuditData, String playerName, int page, Filter filter) {
        int pageSize = config.integer("messages.page-size", 20);
        int start = pageSize * (page - 1);
        int end = Math.min(start + pageSize, reversedAuditData.size());
        int counter = 0;
        String headerFormat = config.string("messages.header-format", "\n<gold>%player_name%'s Audit Data");
        String transactionFormat = config.string("messages.transaction-format", "<hover:<yellow>%comment%><gray>%counter%. <gray>[%date%] <white>%type% <gold>%amount%</hover>");
        String incomeFormat = config.string("messages.income-format", "<green>+%amount%");
        String expenseFormat = config.string("messages.expense-format", "<red>-%amount%");
        String nextPage = config.string("messages.next-page", "<click:run_command:/arc audit %player_name% %filter% %next_page%><hover:show_text:'Click to view next page'><gold>>");
        String prevPage = config.string("messages.prev-page", "<click:run_command:/arc audit %player_name% %filter% %prev_page%><hover:show_text:'Click to view previous page'><gold><");
        String footerFormat = config.string("messages.footer-format", "%prev%<gray>Page <gold>%page% <gray>of <gold>%total_pages%%next%\n");
        List<String> lines = new ArrayList<>();
        lines.add(headerFormat.replace("%player_name%", playerName));
        for (Transaction transaction : reversedAuditData) {
            if (counter >= start && counter < end) {
                String format = transaction.getAmount() > 0 ? incomeFormat : expenseFormat;
                String rounded = String.format("%.2f", Math.abs(transaction.getAmount()));
                String amount = format.replace("%amount%", rounded);
                String counterWithPadding = String.format("%03d", counter + 1);
                String line = transactionFormat
                        .replace("%counter%", counterWithPadding)
                        .replace("%date%", Utils.formatDate(transaction.getTimestamp()))
                        .replace("%type%", transaction.getType().name())
                        .replace("%amount%", amount)
                        .replace("%date2%", Utils.formatDate(transaction.getTimestamp2()))
                        .replace("%comment%", transaction.getComment() == null ? "-" : transaction.getComment());
                lines.add(line);
            }
            counter++;
            if (counter >= end) break;
        }
        int totalPages = (int) Math.ceil((double) reversedAuditData.size() / pageSize);
        String footer = footerFormat
                .replace("%prev%", page > 1 ? prevPage.replace("%player_name%", playerName).replace("%prev_page%", String.valueOf(page - 1)) : "")
                .replace("%page%", String.valueOf(page))
                .replace("%total_pages%", String.valueOf(totalPages))
                .replace("%next%", page < totalPages ? nextPage.replace("%player_name%", playerName).replace("%next_page%", String.valueOf(page + 1)) : "")
                .replace("%filter%", filter == null ? "ALL" : filter.name().toLowerCase());
        lines.add(footer);
        return mm(String.join("\n", lines));
    }

    public static void clear(String player) {
        repo.getOrNull(player.toLowerCase()).thenAccept(auditData -> {
            if (auditData != null) {
                auditData.getTransactions().clear();
                auditData.setDirty(true);
            }
        });
    }

    public static void clearAll() {
        for (AuditData auditData : repo.all()) {
            auditData.getTransactions().clear();
            auditData.setDirty(true);
        }
    }
}

File: src/main/kotlin/ru/arc/audit/AuditManager.kt
package ru.arc.audit

import net.kyori.adventure.audience.Audience
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import ru.arc.ARC
import ru.arc.core.Tasks
import ru.arc.core.modules.EconomyModule
import ru.arc.metrics.MetricsModule
import ru.arc.util.Logging.error
import ru.arc.util.Logging.info
import ru.arc.xserver.playerlist.PlayerManager
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import java.util.concurrent.CompletableFuture
import java.util.concurrent.atomic.AtomicReference

/**
 * Менеджер аудита экономических операций.
 *
 * Facade object for Java interoperability.
 * Delegates to [AuditService] for business logic.
 *
 * For tests, use [AuditService] directly with mocked dependencies.
 */
object AuditManager {

    private lateinit var service: AuditService
    private lateinit var config: AuditConfig
    private var balanceHistoryTask: ru.arc.core.ScheduledTask? = null
    private val migrationReport = AtomicReference<AuditMigrationReport?>()
    private val sessionTracker = EconomySessionTracker()

    private val balanceHistoryPath: Path by lazy {
        ARC.instance.dataPath.resolve("balance-history")
    }

    // ==================== Lifecycle ====================

    /**
     * Initialize with default production dependencies.
     */
    @JvmStatic
    fun init() {
        migrationReport.set(null)
        val scheduler = Tasks.scheduler
        config = AuditConfig.load()

        val monitor = EconomyAuditMonitor(config, registryProvider = MetricsModule::registry)
        val redisStore =
            if (config.storageMode == AuditStorageMode.SQL) {
                null
            } else {
                RedisAuditEventStore(
                    if (ARC.redisManager != null) RedisAuditRepository.create() else InMemoryAuditRepository(),
                )
            }
        val sqlStore =
            if (config.storageMode == AuditStorageMode.REDIS) {
                null
            } else {
                SqlAuditEventStore.open(
                    config = requireNotNull(config.mysql) { "Audit MySQL config is required" },
                    telemetry = AuditStorageTelemetry(MetricsModule::registry, config.storageMode),
                    writerSettings =
                        AuditWriterSettings(
                            batchSize = config.writeBatchSize,
                            flushIntervalMillis = config.writeFlushIntervalMillis,
                            maximumPendingEvents = config.maximumPendingEvents,
                            retryIntervalMillis = config.writeRetryIntervalMillis,
                            shutdownTimeoutSeconds = config.shutdownTimeoutSeconds.toLong(),
                            jobsCoalesceWindowMillis = config.jobsCoalesceWindowSeconds * 1_000L,
                            jobsCoalesceMaximumEvents = config.jobsCoalesceMaximumEvents,
                            jobsCoalescingEnabled = config.storageMode == AuditStorageMode.SQL,
                        ),
                    maintenanceSettings =
                        AuditMaintenanceSettings(
                            enabled = ARC.serverName.equals(config.cleanupOwnerServer, ignoreCase = true),
                            policy = AuditRetentionPolicy(config.retentionDays, config.jobsRawRetentionDays),
                            intervalHours = config.cleanupIntervalHours.toLong(),
                            maxCompactionDaysPerRun = config.maxCompactionDaysPerRun,
                            deleteBatchSize = config.cleanupDeleteBatchSize,
                        ),
                )
            }
        val store =
            when (config.storageMode) {
                AuditStorageMode.REDIS -> requireNotNull(redisStore)
                AuditStorageMode.DUAL -> DualWriteAuditEventStore(requireNotNull(redisStore), requireNotNull(sqlStore))
                AuditStorageMode.SQL -> requireNotNull(sqlStore)
            }
        service = AuditService(
            eventStore = store,
            config = config,
            scheduler = scheduler,
            monitor = monitor,
            localServer = ARC.serverName,
        )

        service.start()
        if (
            config.storageMode == AuditStorageMode.DUAL &&
            ARC.serverName.equals(config.migrationOwnerServer, ignoreCase = true)
        ) {
            AuditLegacyMigration(requireNotNull(redisStore), requireNotNull(sqlStore), config.migrationBatchSize).migrate()
                .whenComplete { report, failure ->
                    if (failure != null) {
                        error("Audit Redis-to-SQL migration failed", failure)
                    } else {
                        migrationReport.set(report)
                        info(
                            "Audit Redis-to-SQL migration finished: scanned={}, inserted={}, duplicates={}, failed={}",
                            report.scanned,
                            report.inserted,
                            report.duplicates,
                            report.failed,
                        )
                    }
                }
        }
        ensureBalanceHistoryDirectory()
        startBalanceHistoryTask(scheduler)
    }

    /**
     * Initialize with custom service (for testing).
     */
    @JvmStatic
    fun init(customService: AuditService, customConfig: AuditConfig = AuditConfig.default()) {
        migrationReport.set(null)
        service = customService
        config = customConfig
        service.start()
    }

    private fun ensureBalanceHistoryDirectory() {
        if (!Files.exists(balanceHistoryPath)) {
            try {
                Files.createDirectories(balanceHistoryPath)
            } catch (e: IOException) {
                error("Failed to create balance history directory", e)
            }
        }
    }

    private fun startBalanceHistoryTask(scheduler: ru.arc.core.TaskScheduler) {
        if (!config.balanceHistoryEnabled) return

        val interval = 20L * 60 * 5 // 5 minutes
        balanceHistoryTask = scheduler.runTimerAsync(20, interval) {
            recordBalanceHistory()
        }
    }

    /**
     * Stop all tasks and cleanup.
     */
    @JvmStatic
    fun cancel() {
        if (::service.isInitialized) {
            service.stop()
        }
        balanceHistoryTask?.cancel()
        balanceHistoryTask = null
        sessionTracker.clear()
        EconomyPendingContextTracker.clear()
        EconomyTransferCorrelationTracker.clear()
        JobsEconomyContextTracker.clear()
    }

    /**
     * Shutdown the repository.
     */
    @JvmStatic
    fun shutdown() {
        cancel()
        if (::service.isInitialized) {
            service.shutdown()
        }
    }

    // ==================== Player Context ====================

    @JvmStatic
    fun join(name: String) {
        service.playerJoined(name)
    }

    @JvmStatic
    fun join(player: Player) {
        service.playerJoined(player.name)
        sessionTracker.joined(player.uniqueId, player.world.name, System.currentTimeMillis())
    }

    @JvmStatic
    fun leave(name: String) {
        service.playerLeft(name)
    }

    @JvmStatic
    fun leave(player: Player) {
        service.playerLeft(player.name)
        sessionTracker.left(player.uniqueId, player.world.name, System.currentTimeMillis())
        EconomyPendingContextTracker.clear(player.uniqueId)
    }

    @JvmStatic
    fun session(playerId: java.util.UUID, currentWorld: String? = null): EconomySessionSnapshot? =
        sessionTracker.snapshot(playerId, currentWorld, System.currentTimeMillis())

    // ==================== Operations ====================

    @JvmStatic
    fun operation(name: String, amount: Double, type: Type, comment: String) {
        service.operation(name, amount, type, comment)
    }

    @JvmStatic
    fun operation(name: String, amount: Double, type: Type, comment: String, metadata: AuditMetadata) {
        service.operation(name, amount, type, comment, metadata)
    }

    @JvmStatic
    fun economyOperation(
        name: String,
        amount: Double,
        type: Type,
        comment: String,
        metadata: AuditMetadata,
        context: EconomyLedgerContext? = null,
    ) {
        service.economyOperation(name, amount, type, comment, metadata, context)
    }

    @JvmStatic
    fun economyAttempt(
        name: String,
        type: Type,
        comment: String,
        metadata: AuditMetadata,
        context: EconomyLedgerContext,
    ) {
        service.economyAttempt(name, type, comment, metadata, context)
    }

    @JvmStatic
    fun unresolvedBalanceSet(
        name: String,
        absoluteBalance: Double,
        metadata: AuditMetadata,
        context: EconomyLedgerContext? = null,
    ) {
        service.unresolvedBalanceSet(name, absoluteBalance, metadata, context)
    }

    @JvmStatic
    fun income(name: String, amount: Double, type: Type, comment: String) {
        service.income(name, amount, type, comment)
    }

    @JvmStatic
    fun expense(name: String, amount: Double, type: Type, comment: String) {
        service.expense(name, amount, type, comment)
    }

    // ==================== Queries ====================

    @JvmStatic
    fun weight(): Long = service.totalWeight()

    @JvmStatic
    fun weightAsync(): CompletableFuture<Long> = service.totalWeightAsync()

    @JvmStatic
    fun storageStatusAsync(): CompletableFuture<AuditStorageStatus> =
        service.storageStatusAsync().thenApply { status -> status.copy(migration = migrationReport.get()) }

    @JvmStatic
    fun economySummary(
        hours: Int,
        limit: Int,
        serverFilter: String? = null,
        shopMaterials: Set<String> = emptySet(),
        concentrationGroups: Map<String, Set<String>> = emptyMap(),
    ): Map<String, Any?> =
        service.economySummary(hours, limit, serverFilter, shopMaterials, concentrationGroups)

    @JvmStatic
    fun economySummaryAsync(
        hours: Int,
        limit: Int,
        serverFilter: String? = null,
        shopMaterials: Set<String> = emptySet(),
        concentrationGroups: Map<String, Set<String>> = emptyMap(),
    ): CompletableFuture<Map<String, Any?>> =
        service.economySummaryAsync(hours, limit, serverFilter, shopMaterials, concentrationGroups)

    @JvmStatic
    fun economySummarySince(
        sinceEpochMs: Long,
        limit: Int,
        serverFilter: String? = null,
        shopMaterials: Set<String> = emptySet(),
        concentrationGroups: Map<String, Set<String>> = emptyMap(),
    ): Map<String, Any?> =
        service.economySummarySince(sinceEpochMs, limit, serverFilter, shopMaterials, concentrationGroups)

    @JvmStatic
    fun economySummarySinceAsync(
        sinceEpochMs: Long,
        limit: Int,
        serverFilter: String? = null,
        shopMaterials: Set<String> = emptySet(),
        concentrationGroups: Map<String, Set<String>> = emptyMap(),
    ): CompletableFuture<Map<String, Any?>> =
        service.economySummarySinceAsync(sinceEpochMs, limit, serverFilter, shopMaterials, concentrationGroups)

    @JvmStatic
    fun sendAudit(audience: Audience, playerName: String, page: Int, filter: AuditFilter) {
        service.sendAudit(audience, playerName, page, filter)
    }

    // ==================== Clear ====================

    @JvmStatic
    fun clear(playerName: String): CompletableFuture<Int> = service.clearPlayerAsync(playerName)

    @JvmStatic
    fun clearAll(): CompletableFuture<Int> = service.clearAllAsync()

    // ==================== Service Access ====================

    /**
     * Get the underlying service for advanced operations.
     */
    fun getService(): AuditService = service

    // ==================== Balance History ====================

    private fun recordBalanceHistory() {
        val economy = EconomyModule.getEconomy() ?: return
        val timestamp = System.currentTimeMillis()

        for (playerName in PlayerManager.getPlayerNames()) {
            try {
                val offlinePlayer = Bukkit.getOfflinePlayer(playerName)
                if (offlinePlayer.name == null || !offlinePlayer.name.equals(playerName, ignoreCase = true)) {
                    error("Failed to get offline player for {}", playerName)
                    continue
                }

                val balance = economy.getBalance(offlinePlayer)
                val playerPath = balanceHistoryPath.resolve("$playerName.csv")

                if (!Files.exists(playerPath)) {
                    Files.createFile(playerPath)
                }

                Files.write(
                    playerPath,
                    "$timestamp,$balance\n".toByteArray(),
                    StandardOpenOption.CREATE,
                    StandardOpenOption.APPEND,
                    StandardOpenOption.WRITE
                )
            } catch (e: IOException) {
                error("Failed to write balance history for {}", playerName, e)
            }
        }
    }

    // ==================== Legacy Compatibility ====================

}
